sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -0,0 +1,361 @@
+                                    CL_CONTEXT_DEVICES, 0, NULL, &deviceCount);
+
+  std::vector<cl_device_id> devicesInCtx(deviceCount);
+
+  ret_err = clGetContextInfo(PI::cast<cl_context>(context), CL_CONTEXT_DEVICES,
+                             deviceCount * sizeof(cl_device_id),
+                             devicesInCtx.data(), NULL);
+
+  if (ret_err != CL_SUCCESS || deviceCount < 1) {
+    if (res_program != nullptr)
+      *res_program = nullptr;
+    return PI::cast<pi_result>(CL_INVALID_CONTEXT);
+  }
+
+  cl_platform_id curPlatform;
+  ret_err = clGetDeviceInfo(devicesInCtx[0], CL_DEVICE_PLATFORM,
+                            sizeof(cl_platform_id), &curPlatform, NULL);
+
+  if (ret_err != CL_SUCCESS) {
+    if (res_program != nullptr)
+      *res_program = nullptr;
+    return PI::cast<pi_result>(CL_INVALID_CONTEXT);
+  }
+
+  size_t devVerSize;
+  ret_err =
+      clGetPlatformInfo(curPlatform, CL_PLATFORM_VERSION, 0, NULL, &devVerSize);
+  std::string devVer(devVerSize, '\0');
+  ret_err = clGetPlatformInfo(curPlatform, CL_PLATFORM_VERSION, devVerSize,
+                              &devVer.front(), NULL);
+
+  if (ret_err != CL_SUCCESS) {
+    if (res_program != nullptr)
+      *res_program = nullptr;
+    return PI::cast<pi_result>(CL_INVALID_CONTEXT);
+  }
+
+  pi_result err = PI_SUCCESS;
+  if (devVer.find("OpenCL 1.0") == std::string::npos &&
+      devVer.find("OpenCL 1.1") == std::string::npos &&
+      devVer.find("OpenCL 1.2") == std::string::npos &&
+      devVer.find("OpenCL 2.0") == std::string::npos) {
+    if (res_program != nullptr)
+      *res_program = PI::cast<pi_program>(clCreateProgramWithIL(
+          PI::cast<cl_context>(context), il, length, PI::cast<cl_int *>(&err)));
+    return err;
+  }
+
+  size_t extSize;
+  ret_err =
+      clGetPlatformInfo(curPlatform, CL_PLATFORM_EXTENSIONS, 0, NULL, &extSize);
+  std::string extStr(extSize, '\0');
+  ret_err = clGetPlatformInfo(curPlatform, CL_PLATFORM_EXTENSIONS, extSize,
+                              &extStr.front(), NULL);
+
+  if (ret_err != CL_SUCCESS ||
+      extStr.find("cl_khr_il_program") == std::string::npos) {
+    if (res_program != nullptr)
+      *res_program = nullptr;
+    return PI::cast<pi_result>(CL_INVALID_CONTEXT);
+  }
+
+  using apiFuncT =
+      cl_program(CL_API_CALL *)(cl_context, const void *, size_t, cl_int *);
+  apiFuncT funcPtr =
+      reinterpret_cast<apiFuncT>(clGetExtensionFunctionAddressForPlatform(
+          curPlatform, "clCreateProgramWithILKHR"));
+
+  assert(funcPtr != nullptr);
+  if (res_program != nullptr)
+    *res_program = PI::cast<pi_program>(funcPtr(
+        PI::cast<cl_context>(context), il, length, PI::cast<cl_int *>(&err)));
+
+  return err;
+}
+
+pi_result OCL(piSamplerCreate)(pi_context context,
+                               const pi_sampler_properties *sampler_properties,
+                               pi_sampler *result_sampler) {
+  // Initialize properties according to OpenCL 2.1 spec.
```

Comment:
Is it the intention of this plugin to take OpenCL 2.1 as "golden" and map all interface to this version of OpenCL? 
It is fine, but it should be made clearer maybe at the top or in the plugin documentation

Reply:
@smaslov-intel Please comment on this.